File: Source/Map/Cell.cpp
#include <Cell.h>

#include <Game.h>
#include <Map.h>
#include <Face.h>
#include <Actor.h>
#include <DataManager.h>

#include <Renderer.h>
#include <TextureManager.h>
#include <TerrainRender.h>


Cell::Cell()
{
    CubeShape EmptyCube = CubeShape(BELOW_CUBE_HEIGHT);

    CubeShapeTypes.assign(EmptyCube);
    CubeMaterialTypes.assign(INVALID_INDEX);

    Hidden.reset();
    SubTerranean.reset();
    SkyView.reset();
    SunLit.reset();

	Render = NULL;
}

Cell::~Cell()
{

}

void Cell::setCellPosition(CellCoordinates Coordinates)
{
    thisCellCoordinates = Coordinates;
    if (Render == NULL)
	{
		Render = new TerrainRendering(this);
	} else {
		//Render->
	}
}

void Cell::setCubeShape(CubeCoordinates Coordinates, CubeShape NewShape)
{
    if (NewShape != CubeShapeTypes[Coordinates])
    {
        CubeShapeTypes[Coordinates] = NewShape;
        Render->setDirty();

        Face* TargetFace = getFace(FaceCoordinates(Coordinates, DIRECTION_NONE));
        if (TargetFace != NULL)
        {
            setFaceShape(FaceCoordinates(Coordinates, DIRECTION_NONE), FaceShape(NewShape, DIRECTION_NONE));
        }
    }
}

void Cell::BuildFaceData()
{
	Map* ParentMap = GAME->getMap();
    MapCoordinates TargetMapCoordinates;
    bool Debug = true;

    CubeCoordinates TargetCube = 0;
    do
    {
        CubeShape Shape = getCubeShape(TargetCube);
        int16_t CubeMaterial = getCubeMaterial(TargetCube);

        //static int16_t NEHEMaterial = DATA->getLabelIndex("MATERIAL_UNINITIALIZED");
        static int16_t WallSurface = DATA->getLabelIndex("SURFACETYPE_ROUGH_WALL");
        static int16_t FloorSurface = DATA->getLabelIndex("SURFACETYPE_ROUGH_FLOOR_1");

        for (uint8_t i = 0, DirectionType = Direction::AXIAL_DIRECTIONS[i]; i < NUM_AXIAL_DIRECTIONS; ++i, DirectionType = Direction::AXIAL_DIRECTIONS[i])
        {
            FaceCoordinates FaceLocation = FaceCoordinates(TargetCube, DirectionType);
            MapCoordinates ModifiedCoordinates = MapCoordinates(thisCellCoordinates, TargetCube);
            ModifiedCoordinates.TranslateMapCoordinates(DirectionType);

            CubeShape AdjacentShape = ParentMap->getCubeShape(ModifiedCoordinates);

            if (!Shape.isSky())
            {
                if (ParentMap->isCubeInited(ModifiedCoordinates) && AdjacentShape.isSky())
                {
                    if (Shape.hasFace(DirectionType))
                    {
                        Face* NewFace = ParentMap->addFace(MapCoordinates(thisCellCoordinates, TargetCube), DirectionType);

                        NewFace->setFaceMaterialType(CubeMaterial);
                        NewFace->setFaceSurfaceType(WallSurface);
                        NewFace->setFaceShapeType(FaceShape(Shape, DirectionType));
                    }
                }

                if (!AdjacentShape.isEmpty())
                {
                    if (DirectionType == DIRECTION_DOWN && Shape.hasFloor())
                    {
                        Face* NewFace = ParentMap->addFace(MapCoordinates(thisCellCoordinates, TargetCube), DirectionType);

                        NewFace->setFaceMaterialType(ParentMap->getCubeMaterial(ModifiedCoordinates));
                        NewFace->setFaceSurfaceType(FloorSurface);
                        NewFace->setFaceShapeType(FaceShape(Shape, DirectionType));
                    }
                }
            }
        }

        if (!Shape.isEmpty() && !Shape.isSolid())
        {
            Face* NewFace = addFace(FaceCoordinates(TargetCube, DIRECTION_NONE));

            NewFace->setFaceMaterialType(CubeMaterial);
            NewFace->setFaceSurfaceType(FloorSurface);
            NewFace->setFaceShapeType(FaceShape(Shape, DIRECTION_NONE));
        }

        TargetCube++;

    }
    while (TargetCube != 0);  // End Loop when Byte rolls over
	Render->setDirty();
}

void Cell::RegisterWithRendering()
{
	Render->RegisterWithRendering();
}

int16_t Cell::getFaceMaterialType(FaceCoordinates TargetCoordinates) const
{
    Face* TargetFace = getFace(TargetCoordinates);
    return TargetFace != NULL ? TargetFace->getFaceMaterialType() : INVALID_INDEX;
}

int16_t Cell::getFaceSurfaceType(FaceCoordinates TargetCoordinates) const
{
    Face* TargetFace = getFace(TargetCoordinates);
    return TargetFace != NULL ? TargetFace->getFaceSurfaceType() : INVALID_INDEX;
}

bool Cell::setFaceMaterialType(FaceCoordinates TargetCoordinates, int16_t MaterialTypeID)
{
    Face* TargetFace = getFace(TargetCoordinates);

    if (TargetFace != NULL)
    {
        TargetFace->setFaceMaterialType(MaterialTypeID);
        Render->setDirty();
        return true;
    }
    return false;
}

bool Cell::setFaceSurfaceType(FaceCoordinates TargetCoordinates, int16_t SurfaceTypeID)
{
    Face* TargetFace = getFace(TargetCoordinates);

    if (TargetFace != NULL)
    {
        TargetFace->setFaceSurfaceType(SurfaceTypeID);
        Render->setDirty();
        return true;
    }
    return false;
}

FaceShape Cell::getFaceShape(FaceCoordinates TargetCoordinates) const
{
    Face* TargetFace = getFace(TargetCoordinates);
    return TargetFace != NULL ? TargetFace->getFaceShapeType() : FaceShape(CubeShape(false), DIRECTION_NONE);
}

bool Cell::setFaceShape(FaceCoordinates TargetCoordinates, FaceShape NewShape)
{
    Face* TargetFace = getFace(TargetCoordinates);

    if (TargetFace != NULL)
    {
        TargetFace->setFaceShapeType(NewShape);
        Render->setDirty();
        return true;
    }
    return false;
}

bool Cell::removeFace(FaceCoordinates TargetCoordinates)
{
    uint16_t Key = TargetCoordinates.FaceKey();

    if (Faces.find(Key) != Faces.end())
    {
        delete Faces.find(Key)->second;
        Faces.erase(Key);

        Render->setDirty();
        return true;
    }
    return false;
}

Face* Cell::addFace(FaceCoordinates TargetCoordinates)
{
    uint16_t Key = TargetCoordinates.FaceKey();

    if (Faces.find(Key) == Faces.end())
    {
        Face* NewFace = new Face(TargetCoordinates.Coordinates, TargetCoordinates.FaceDirection);
        Faces[Key] = NewFace;

        return NewFace;
    }
    else
    {
        return Faces.find(Key)->second;
    }
}

void Cell::addActor(Actor* NewActor)
{
	Render->AddActor(NewActor);
}

void Cell::removeActor(Actor* DepartingActor)
{
	Render->RemoveActor(DepartingActor);
}

void Cell::setNeedsReRendering()
{
	Render->setDirty();
}

void Cell::Save(boost::filesystem::basic_ofstream<char>& Stream) const
{
	Stream.write((char*)&thisCellCoordinates.X, sizeof(thisCellCoordinates.X));
	Stream.write((char*)&thisCellCoordinates.Y, sizeof(thisCellCoordinates.Y));
	Stream.write((char*)&thisCellCoordinates.Z, sizeof(thisCellCoordinates.Z));

	for (int i = 0; i < CUBESPERCELL; i++)
	{
		Stream.write((char*)&CubeShapeTypes[i], sizeof(CubeShapeTypes[i]));
		Stream.write((char*)&CubeMaterialTypes[i], sizeof(CubeMaterialTypes[i]));
	}

	int16_t FaceCount = Faces.size();
	Stream.write((char*)&FaceCount, sizeof(FaceCount));
	for (std::map<uint16_t, Face*>::const_iterator it = Faces.begin(); it != Faces.end(); it++)
	{
		uint16_t Key = it->first;
		Stream.write((char*)&Key, sizeof(Key));
		it->second->Save(Stream);
	}
}

void Cell::Load(boost::filesystem::basic_ifstream<char>& Stream)
{
	Stream.read((char*)&thisCellCoordinates.X, sizeof(thisCellCoordinates.X));
	Stream.read((char*)&thisCellCoordinates.Y, sizeof(thisCellCoordinates.Y));
	Stream.read((char*)&thisCellCoordinates.Z, sizeof(thisCellCoordinates.Z));

	for (int i = 0; i < CUBESPERCELL; i++)
	{
		Stream.read((char*)&CubeShapeTypes[i], sizeof(CubeShapeTypes[i]));
		Stream.read((char*)&CubeMaterialTypes[i], sizeof(CubeMaterialTypes[i]));
	}

	int16_t FaceCount;
	Stream.read((char*)&FaceCount, sizeof(FaceCount));
	for (int i = 0; i < FaceCount; i++)
	{
		uint16_t Key;
		Stream.read((char*)&Key, sizeof(Key));
		Face* NewFace = new Face();
		NewFace->Load(Stream);
		Faces[Key] = NewFace;
	}

	Render = new TerrainRendering(this);
}

File: src/Map/Cell.java
/* Copyright 2010 Kenneth 'Impaler' Ferland

 This file is part of Khazad.

 Khazad is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Khazad is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Khazad.  If not, see <http://www.gnu.org/licenses/> */

package Map;

import Data.DataManager;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.Serializable;

import java.util.HashMap;
import java.util.BitSet;

import org.javatuples.Pair;

/**
 * Primary MapData storage class, holds all data for describing a 16x16 square
 * 1 level high region of the map, each tile is indexed by a byte for array and
 * bitset based data and HashMaps hold face objects.
 *
 * @author Impaler
 */
public class Cell implements Serializable {

	private static final long serialVersionUID = 1;
	// Larger DataValues specific to each Cube
	private short[] CubeMaterialTypes;
	private short[][] CubeShapeTypes;
	// Bit values for each Cube
	private BitSet Hidden;
	private BitSet SubTerranean;
	private BitSet SkyView;
	private BitSet SunLit;
	// Keeps all Faces between and inside Cubes
	private HashMap<FaceCoordinate, Face>[] Faces;
	// The global position of this cell relative to other cells
	private CellCoordinate thisCellCoordinates;
	// Dirty values, set true on changes, set false by rendering
	transient boolean DirtyTerrainRendering;
	transient boolean DirtyPathRendering;

	protected static final short WallSurface = DataManager.getLabelIndex("SURFACETYPE_ROUGH_WALL");
	protected static final short FloorSurface = DataManager.getLabelIndex("SURFACETYPE_ROUGH_FLOOR_1");

	String CellString;

	public Cell() {
		CubeMaterialTypes = new short[CubeCoordinate.CUBESPERCELL];
		CubeShapeTypes = new short[CubeCoordinate.CELLDETAILLEVELS][];
		Faces = (HashMap<FaceCoordinate, Face>[]) new HashMap<?, ?>[CubeCoordinate.CELLDETAILLEVELS];

		for (int i = 0; i < CubeCoordinate.CELLDETAILLEVELS; i++) {
			Faces[i] = new HashMap<FaceCoordinate, Face>();
			int Size = (CubeCoordinate.CELLDETAILLEVELS - i) - 1;
			Size = 1 << Size;
			CubeShapeTypes[i] = new short[Size * Size];
			for (int j = 0; j < CubeShapeTypes[i].length; j++) {
				CubeShapeTypes[i][j] = CubeShape.EMPTY_CUBE_DATA;
			}
		}

		Hidden = new BitSet(CubeCoordinate.CUBESPERCELL);
		SubTerranean = new BitSet(CubeCoordinate.CUBESPERCELL);
		SkyView = new BitSet(CubeCoordinate.CUBESPERCELL);
		SunLit = new BitSet(CubeCoordinate.CUBESPERCELL);

		for (int i = 0; i < CubeMaterialTypes.length; i++) {
			CubeMaterialTypes[i] = DataManager.INVALID_INDEX;
		}

		DirtyTerrainRendering = true;
		DirtyPathRendering = true;
	}

	private void readObject(ObjectInputStream ois) throws ClassNotFoundException, IOException {
		// default deserialization
		ois.defaultReadObject();
		// fix transients
		DirtyTerrainRendering = true;
		DirtyPathRendering = true;
	}

	public void setCellCoordinates(CellCoordinate Coordinates) {
		thisCellCoordinates = Coordinates;
		CellString = getClass().getName() + thisCellCoordinates.toString();
		setRenderingDirty();
	}

	public CellCoordinate getCellCoordinates() {
		return thisCellCoordinates.clone();
	}

	public void setCubeShape(short Coordinates, CubeShape NewShape, int Level) {
		if (NewShape.getData() != CubeShapeTypes[Level][Coordinates]) {
			CubeShapeTypes[Level][Coordinates] = NewShape.getData();

			Face TargetFace = getFace(new FaceCoordinate(Coordinates, Direction.DIRECTION_NONE), 0);
			if (TargetFace != null) {
				setFaceShape(new FaceCoordinate(Coordinates, Direction.DIRECTION_NONE), new FaceShape(NewShape, null, Direction.DIRECTION_NONE));
			}
			setRenderingDirty();
		}
	}

	public void setCubeShape(short Coordinates, CubeShape NewShape) {
		if (NewShape.getData() != CubeShapeTypes[0][Coordinates]) {
			CubeShapeTypes[0][Coordinates] = NewShape.getData();

			Face TargetFace = getFace(new FaceCoordinate(Coordinates, Direction.DIRECTION_NONE), 0);
			if (TargetFace != null) {
				setFaceShape(new FaceCoordinate(Coordinates, Direction.DIRECTION_NONE), new FaceShape(NewShape, null, Direction.DIRECTION_NONE));
			}
			setRenderingDirty();
		}	
	}
	
	public void buildFaces(int LevelofDetail) {
		GameMap ParentMap = GameMap.getMap();
	
		int SizeFactor = (CubeCoordinate.CELLDETAILLEVELS - LevelofDetail) - 1;
		int Size = 1 << SizeFactor;
		int Count = Size * Size;
		for (short j = 0; j < Count; j++) {
			CubeShape Shape = getCubeShape(j, LevelofDetail);
			short CubeMaterial = getCubeMaterial(j);

			for (Direction DirectionType : Direction.AXIAL_DIRECTIONS) {
				Pair <CellCoordinate, Short> values = ParentMap.TranslateCubeIndex(thisCellCoordinates, j, DirectionType, LevelofDetail);				
				CubeCoordinate AdjacentCoordinates = new CubeCoordinate(values.getValue0(), values.getValue1().shortValue());

				if (ParentMap.isCubeInited(AdjacentCoordinates)) {
					CubeShape AdjacentShape = ParentMap.getCubeShape(AdjacentCoordinates);

					if (AdjacentShape.isSky()) {
						if (Shape.hasFace(DirectionType)) {
							Face NewFace = ParentMap.addFace(new CubeCoordinate(thisCellCoordinates, j), DirectionType, LevelofDetail);

							NewFace.setFaceMaterialType(CubeMaterial);
							NewFace.setFaceSurfaceType(WallSurface);
							NewFace.setFaceShapeType(new FaceShape(Shape, AdjacentShape, DirectionType));
							if (ParentMap.isCubeSunLit(AdjacentCoordinates))
								NewFace.Sunlit = true;
						}
					}

					if (!AdjacentShape.isEmpty()) {
						if (DirectionType == Direction.DIRECTION_DOWN && Shape.hasFloor() && AdjacentShape.hasCeiling()) {
							Face NewFace = ParentMap.addFace(new CubeCoordinate(thisCellCoordinates, j), DirectionType, LevelofDetail);

							NewFace.setFaceMaterialType(ParentMap.getCubeMaterial(AdjacentCoordinates));
							NewFace.setFaceSurfaceType(FloorSurface);
							NewFace.setFaceShapeType(new FaceShape(Shape, AdjacentShape, DirectionType));
							if (ParentMap.isCubeSunLit(AdjacentCoordinates))
								NewFace.Sunlit = true;
						}
					}
				}
			}

			if (!Shape.isEmpty() && !Shape.isSolid()) {
				Face NewFace = addFace(new FaceCoordinate(j, Direction.DIRECTION_NONE, LevelofDetail), LevelofDetail);

				NewFace.setFaceMaterialType(CubeMaterial);
				NewFace.setFaceSurfaceType(FloorSurface);
				NewFace.setFaceShapeType(new FaceShape(Shape, null, Direction.DIRECTION_NONE));
				if (isCubeSunLit(j))
					NewFace.Sunlit = true;
			}
		}
		setRenderingDirty();
	}

	public void growGrass() {
		DataManager Data = DataManager.getDataManager();

		for (Face TargetFace : Faces[0].values()) {
			if (TargetFace.Sunlit) {
				short MaterialID = TargetFace.getFaceMaterialType();
				int GrowthFactor = Data.getMaterialData(MaterialID).PlantGrowthFactor;
				if (GrowthFactor > 0) {
					TargetFace.setFaceMaterialType(Data.getLabelIndex("MATERIAL_DARK_GRASS"));
				}
			}
		}
	}

	public Face getFace(FaceCoordinate TargetCoordinates, int LevelofDetail) {
		return Faces[LevelofDetail].get(TargetCoordinates);
	}

	boolean hasFace(FaceCoordinate TargetCoordinates, int LevelofDetail) {
		return Faces[LevelofDetail].containsKey(TargetCoordinates);
	}

	short getFaceMaterialType(FaceCoordinate TargetCoordinates) {
		Face TargetFace = getFace(TargetCoordinates, 0);
		return TargetFace != null ? TargetFace.getFaceMaterialType() : DataManager.INVALID_INDEX;
	}

	short getFaceSurfaceType(FaceCoordinate TargetCoordinates) {
		Face TargetFace = getFace(TargetCoordinates, 0);
		return TargetFace != null ? TargetFace.getFaceSurfaceType() : DataManager.INVALID_INDEX;
	}

	boolean setFaceMaterialType(FaceCoordinate TargetCoordinates, short MaterialTypeID) {
		Face TargetFace = getFace(TargetCoordinates, 0);

		if (TargetFace != null) {
			TargetFace.setFaceMaterialType(MaterialTypeID);
			setDirtyTerrainRendering(true);
			return true;
		}
		return false;
	}

	boolean setFaceSurfaceType(FaceCoordinate TargetCoordinates, short SurfaceTypeID) {
		Face TargetFace = getFace(TargetCoordinates, 0);

		if (TargetFace != null) {
			TargetFace.setFaceSurfaceType(SurfaceTypeID);
			setDirtyTerrainRendering(true);
			return true;
		}
		return false;
	}

	FaceShape getFaceShape(FaceCoordinate TargetCoordinates) {
		Face TargetFace = getFace(TargetCoordinates, 0);
		return TargetFace != null ? TargetFace.getFaceShapeType() : new FaceShape(new CubeShape(), new CubeShape(), Direction.DIRECTION_NONE);
	}

	boolean setFaceShape(FaceCoordinate TargetCoordinates, FaceShape NewShape) {
		Face TargetFace = getFace(TargetCoordinates, 0);

		if (TargetFace != null) {
			TargetFace.setFaceShapeType(NewShape);
			setRenderingDirty();
			return true;
		}
		return false;
	}

	boolean removeFace(FaceCoordinate TargetCoordinates, int LevelofDetail) {
		if (Faces[LevelofDetail].containsKey(TargetCoordinates)) {
			Faces[LevelofDetail].remove(TargetCoordinates);
			setRenderingDirty();
			return true;
		}
		return false;
	}

	Face addFace(FaceCoordinate TargetCoordinates, int LevelofDetail) {
		Face TargetFace = Faces[LevelofDetail].get(TargetCoordinates);
		if (TargetFace == null) {
			Face NewFace = new Face(TargetCoordinates.getCoordinates(), TargetCoordinates.getFaceDirection());
			Faces[LevelofDetail].put(TargetCoordinates, NewFace);
			setRenderingDirty();
			return NewFace;
		} else {
			return TargetFace;
		}
	}

	public HashMap<FaceCoordinate, Face> getFaces(int LevelofDetail) {
		return Faces[LevelofDetail];
	}

	public void setRenderingDirty() {
		setDirtyTerrainRendering(true);
		setDirtyPathingRendering(true);
	}

	public void setDirtyTerrainRendering(boolean dirty) {
		DirtyTerrainRendering = dirty;
	}

	public void setDirtyPathingRendering(boolean dirty) {
		DirtyPathRendering = dirty;
	}

	public boolean isTerrainRenderingDirty() {
		return DirtyTerrainRendering;
	}

	public boolean isPathingRenderingDirty() {
		return DirtyPathRendering;
	}

	public CubeShape getCubeShape(short Index, int LevelofDetal) {
		return new CubeShape(CubeShapeTypes[LevelofDetal][Index]);
	}

	public short getCubeMaterial(short Coordinates) {
		return CubeMaterialTypes[Coordinates];
	}

	public void setCubeMaterial(short Coordinates, short MaterialID) {
		CubeMaterialTypes[Coordinates] = MaterialID;
		DirtyTerrainRendering = true;
	}

	public boolean isCubeHidden(short Coordinates) {
		return Hidden.get(Coordinates);
	}

	public void setCubeHidden(short Coordinates, boolean NewValue) {
		Hidden.set(Coordinates, NewValue);
		DirtyTerrainRendering = true;
	}

	public boolean isCubeSubTerranean(short Coordinates) {
		return SubTerranean.get(Coordinates);
	}

	public void setCubeSubTerranean(short Coordinates, boolean NewValue) {
		SubTerranean.set(Coordinates, NewValue);
		DirtyTerrainRendering = true;
	}

	public boolean isCubeSkyView(short Coordinates) {
		return SkyView.get(Coordinates);
	}

	public void setCubeSkyView(short Coordinates, boolean NewValue) {
		SkyView.set(Coordinates, NewValue);
		DirtyTerrainRendering = true;
	}

	public boolean isCubeSunLit(short Coordinates) {
		return SunLit.get(Coordinates);
	}

	public void setCubeSunLit(short Coordinates, boolean NewValue) {
		SunLit.set(Coordinates, NewValue);
		DirtyTerrainRendering = true;
	}

	@Override
	public String toString() {
		return CellString;
	}
}
